### PR TITLE
Feature/dummy revision

### PR DESCRIPTION
**Pull request by lianxmfor**
This PR does the following:

1. oomstore adds a new method CreateRevision, which checks and creates dummy revision
2. automatically create a domain revision when the group creates its first revisions

### CODE REVIEW

**Comment by wfxr** [comment]
Did you mean dummy revision ?

**Comment by lianxmfor** [comment]
> 

I heard it as `domain` 😱

**Comment by lianxmfor** [comment]
> Did you mean dummy revision ?

I heard it as domain 😱

File: pkg/oomstore/revision.go [hunk 1]
```diff
@@ -24,6 +25,50 @@ func (s *OomStore) GetRevisionBy(ctx context.Context, groupID int, revision int6
...
 }
 
+func (s *OomStore) CreateRevision(ctx context.Context, opt metadata.CreateRevisionOpt) (int, string, error) {
+	var (
+		revisionID    int
+		snapshotTable string
+		err           error
+
+		domainRevision *types.Revision
+	)
```
Comment (wfxr) [inline pkg/oomstore/revision.go:35]: Can you define the variables near to it's uage if possible? It's hard to follow what the  value was for each variable and when it was used in this style.

Comment (lianxmfor) [reply pkg/oomstore/revision.go:35]: These variables will be assigned values in the WithTransaction and can only be defined outside.
Another way to write it is:

```go
var revisionID int
var snapshotTable string
var err error
var domainRevision *types.Revision
```

Comment (wfxr) [reply pkg/oomstore/revision.go:35]: Ok. err is not nesessary pre decleared?
```go
var revisionID int
var snapshotTable string
var domainRevision *types.Revision

if err := ...
```

Comment (lianxmfor) [reply pkg/oomstore/revision.go:35]: err can be defined where it is used, I'll change it later. 

```
var revisionID int
var snapshotTable string
var domainRevision *types.Revision
```
is better than?

```
var (
    revisionID int
    snapshotTable string
    domainRevision *types.Revision
)
```

Comment (wfxr) [reply pkg/oomstore/revision.go:35]: I prefer the first but it's ok if you insist the second.

File: pkg/oomstore/revision.go [hunk 2]
```diff
@@ -24,6 +25,46 @@ func (s *OomStore) GetRevisionBy(ctx context.Context, groupID int, revision int6
...
+		_, err := s.metadata.GetRevisionBy(ctx, opt.GroupID, 0)
+		if err != nil {
+			if !errdefs.IsNotFound(err) {
+				return err
+			}
+
+			if _, _, err = s.metadata.CreateRevision(ctx, metadata.CreateRevisionOpt{
+				Revision:    0,
+				GroupID:     opt.GroupID,
+				Description: "dummy revision it is useful for join and export",
```
Comment (jinghancc) [inline pkg/oomstore/revision.go:43]: suggestion: dummy revision will be used at Join and Export

Comment (lianxmfor) [reply pkg/oomstore/revision.go]: done

**Review by wfxr — APPROVED** [review]
LGTM

File: pkg/oomstore/revision.go [hunk 3]
```diff
@@ -24,6 +25,46 @@ func (s *OomStore) GetRevisionBy(ctx context.Context, groupID int, revision int6
...
+			}
+		}
+
+		revisionID, snapshotTable, err = s.metadata.CreateRevision(ctx, opt)
+		return err
+	}); err != nil {
+		return 0, "", err
+	}
+
+	if dummyRevision != nil {
```
Comment (jinghancc) [inline pkg/oomstore/revision.go:59]: can you move the logic right after line 51?

Comment (jinghancc) [reply pkg/oomstore/revision.go:59]: also, `createFirstSnapshotTable` needs to be deprecated; for dummy revision, we need to create both snapshot table and cdc table

Comment (lianxmfor) [reply pkg/oomstore/revision.go:59]: - can you move the logic right after line 51?

Moving after line 51 causes the offline store SQL to be executed in the meta store transaction, which would be a problem in SQLite, so it's written here to avoid that, similar to apply's onlineJobs.

Comment (jinghancc) [reply pkg/oomstore/revision.go:59]: ah I see!

Comment (lianxmfor) [reply pkg/oomstore/revision.go:59]: - also, createFirstSnapshotTable needs to be deprecated; for dummy revision, we need to create both snapshot table and cdc table

I created a new function: `createSnapshotAndCdcTable` instead of `createFirstSnapshotTable`

File: pkg/oomstore/revision.go [hunk 4]
```diff
@@ -24,6 +25,85 @@ func (s *OomStore) GetRevisionBy(ctx context.Context, groupID int, revision int6
...
+		Features:  features,
+		TableType: types.TableStreamSnapshot,
+	}); err != nil {
+		return err
+	}
+
+	// Update snapshot_table in feature_group_revision table
+	if err := s.metadata.UpdateRevision(ctx, metadata.UpdateRevisionOpt{
+		RevisionID:       revision.ID,
+		NewSnapshotTable: &snapshotTable,
```
Comment (jinghancc) [inline pkg/oomstore/revision.go:88]: should we also set `CdcTable` for the revision?

**Review by jinghancc — APPROVED** [review]
LGTM

File: pkg/oomstore/revision.go (hunk 4 above)
Comment (lianxmfor) [reply pkg/oomstore/revision.go:91]: `metadata.UpdateRevision` currently does not support updating the `CdcTable`, I want to change it in the next PR.

Comment (jinghancc) [reply pkg/oomstore/revision.go:91]: sounds good!
